src/sonic-framework/rebootbackend/reboot_thread.h

```diff
@@ -0,0 +1,222 @@
+//   called from the same thread.
+class RebootThread {
+ public:
+  enum class Status { SUCCESS, FAILURE, KEEP_WAITING };
+  enum class Progress { PROCEED, EXIT_EARLY };
+
+  // interface: dbus reboot host service access
+  // m_finished: let launching task know thread has finished
+  RebootThread(DbusInterface &dbus_interface,
+               swss::SelectableEvent &m_finished);
+
+  NotificationResponse Start(const gnoi::system::RebootRequest &request);
+
+  // Request thread stop/exit. Only used when platform is shutting down
+  // all containers/processes.
+  void Stop(void);
+
+  // Called by launching task after notification sent to m_finished.
+  bool Join(void);
+
+  // Return Status of last reboot attempt
+  gnoi::system::RebootStatusResponse GetResponse();
+
+  // Returns true if the RebootThread has been started since the last reboot,
+  // and false otherwise.
+  bool HasRun();
+
+ private:
+  void reboot_thread(void);
+  void do_reboot(void);
+  Progress send_dbus_reboot_request();
+  void do_cold_reboot(swss::Select &s);
+
+  // Inner loop select handler to wait for platform reboot.
+  //   wait for timeout
+  //   wait for a stop request (sigterm)
+  // Returns:
+  //   EXIT_EARLY: an issue occurred that stops WARM
+  //   PROCEED: if reboot timeout expired
+  Progress platform_reboot_select(swss::Select &s,
+                                  swss::SelectableTimer &l_timer);
+
+  // Wait for platform to reboot while waiting for possible stop
+  // Returns:
+  //   EXIT_EARLY: an issue occurred that stops WARM
+  //   PROCEED: if reboot timeout expired
+  Progress wait_for_platform_reboot(swss::Select &s);
+
+  // Log error string, set status to RebootStatus_Status_STATUS_FAILURE
+  // Set status message to error_string.
+  void log_error_and_set_non_retry_failure(const std::string error_string);
+
+  // Log error string, set status to
+  // RebootStatus_Status_STATUS_RETRIABLE_FAILURE Set status message to
+  // error_string.
+  void log_error_and_set_failure_as_retriable(const std::string error_string);
+
+
+  // Request is input only.
+  // Response is ouput only.
+  // Return true if preconditions met, false otherwise.
+  bool check_start_preconditions(const gnoi::system::RebootRequest &request,
+                                 NotificationResponse &response);
+  std::thread m_thread;
+
+  // Signal m_finished to let main thread know weve completed.
+  // Main thread should call Join.
+  swss::SelectableEvent &m_finished;
+
+  // m_stop signalled by main thread on sigterm: cleanup and exit.
+  swss::SelectableEvent m_stop;
+  DbusInterface &m_dbus_interface;
+  swss::DBConnector m_db;
+  ThreadStatus m_status;
+  gnoi::system::RebootRequest m_request;
+
+  // Wait for system to reboot: allow unit test to shorten.
+  // TODO: there is a plan to make these timer values
+  //       available in CONFIG_DB
+  static constexpr uint32_t kRebootTime = 260;
+  long m_reboot_timeout = kRebootTime;
```

Comment:
If we want to change this for testing purposes, we should set this value at constructor. Friend is in general considered dangerous and bad for encapsulation.

Reply:
I think this is debatable?
Google coding standards allow this: “In some cases it may be useful to make a unittest class a friend of the class it tests.
This timeout is adjusted on a test by test basis: 
The unit test could no longer be globally instantiated, and would have to be nstantiated  per test.
Fair amount of rework to support this.